### DISCUSSION TITLE
Does AtomFamily always create a new map entry?

### DISCUSSION THREAD

**Discussion by tmfalves** [pmndrs/jotai#946] [Q&A]
Hello everyone! First of all, great project, I've mainly been using Redux and React's own state management and it's been a very interesting shift to Jotai.

I've been implementing some atomFamily states and something that struck me is that regardless of the situation an atom is always created for a given parameter. In my understanding, wouldn't it make sense for it to allow something like a "find" without adding a new entry to the map? Or am I missing the point of atomFamily completely?

**Comment by dai-shi** [comment]
`atomFamily` will create a `Map`, so it a has to be called just once.
Also, the default equality function is like `===`, which is customizable with `areEqual`.

Other than that, it's hard to tell something. You may want to provide some code snippets.
Alternatively, you can try reading the `atomFamily` code, the main part is fairly small:
https://github.com/pmndrs/jotai/blob/f8669dc43b27c8cea82c2d01eb0770746ac95832/src/utils/atomFamily.ts#L16-L36

All that said, `atomFamily` is developed more or less for migrating from Recoil.
In the "Jotai" way, we don't need to use `atomFamily`. Atoms can be created pretty easily without string keys, and can be GC'd. (With `atomFamily`, developers are responsible to clean up to avoid memory leaks.)

**Reply by tmfalves** [reply]
Coming from Recoil, so I'm probably still in that mindset. You're right, a snippet might illustrate my predicament better, basically I have an application with some objects and this objects can have components (these can be different depending on the object), what I'm doing right now is something like this to add a new object


```typescript
export const addGameObject = atom<null, { id: string; type: OBJECT_TYPES }>(
  null,
  (get, set, newValue) => {
    const { id } = newValue;

    set(gameObjectsState, [
      ...get(gameObjectsState),
      { id, type: newValue.type, meta: {}, status: STATUS.IDLE },
    ]);

    gameObjectsPositions({ id });
    gameObjectsScale({ id });

    infoAtomFamily(id);
    transformAtomFamily(id);

    if (newValue.type !== OBJECT_TYPES.EMPTY) {
      rendererAtomFamily(id);
      colliderAtomFamily(id);
      rigidbodyAtomFamily(id);
    }
  }
);
```

So some objects don't necessarily have a component initiated in a certain family. However if I try to just "read", say, the `rendererAtomFamily` (in my head I was expecting this to return undefined for example)

```typescript
  const values = useAtomValue(rendererAtomFamily(props.id));
```

it will always create the atom with defaults

```typescript
export const rendererAtomFamily = atomFamily((id: string) =>
  atom({
    objId: id,
    ...DEFAULT_RENDER_PARAMS,
  })
);
```

So I'm assuming this is intended and I'm probably misusing `atomFamily`?

**Reply by dai-shi** [reply]
Hm, interesting. This actually indicates that I might be misunderstanding how `atomFamily` works in recoil.
Anyhow, `atomFamily` in jotai is a simple Map. So, when you first use it:
```js
rendererAtomFamily(props.id)
```
it will create a new atom, store it in the Map, and return it. And then, for the second time, you will get a cached atom. It will never return `undefined`.

And this means, there's no point of doing `rendererAtomFamily(id);` in addGameObject's write function.

**Reply by tmfalves** [reply]
Most likely you translated the `atomFamily` correctly from Recoil, I might be using it in a way that actually requires a different strategy altogether either way, thank you for clearing it up, `atomFamily` will always store a new atom when called (first time at least) and as you said no need for that call on `addGameObject`. I'll try the atom in atom pattern I've seen you suggesting in other discussions, seems more appropriate in this particular case.

Thank you for your support!

**Reply by dai-shi** [reply]
Great! Feel free to share your thoughts when you make some progress.
